File: fixtures/scale/scala/large/src/cryptobench/ciphers/sdbm/SdbmHash.scala
package cryptobench.ciphers.sdbm

import cryptobench.core.HashFunction
import cryptobench.util.Bytes

/** sdbm hash as used by the old sdbm database library. */
final class SdbmHash extends HashFunction {

  override def name: String = "sdbm"

  override def digest(input: String): Long = {
    var state = 0L
    for (b <- Bytes.of(input)) {
      state = (b.toLong & 0xFF) + (state << 6) + (state << 16) - state
    }
    state
  }
}

File: fixtures/scale/dart/large/lib/ciphers/sdbm/sdbm_hash.dart
import '../../core/hash_function.dart';
import '../../util/bytes.dart';

/// sdbm hash as used by the old sdbm database library.
class SdbmHash implements HashFunction {
  @override
  String name() => 'sdbm';

  @override
  int digest(String input) {
    var state = 0;
    for (final b in bytesOf(input)) {
      state = (b + (state << 6) + (state << 16) - state) & 0xFFFFFFFF;
    }
    return state;
  }
}
